File: src/snap.c
/*
 * Copyright (C) 2015 Canonical Ltd
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "config.h"
#include "snap.h"

#include <stddef.h>
#include <stdlib.h>
#include <regex.h>

#include "utils.h"

bool verify_appname(const char *appname)
{
	// snappy appname is of form:
	// snap.<name>.<app>
	// - <name> must start with lowercase letter, then may contain
	//   lowercase alphanumerics and '-'
	// - <app> may contain alphanumerics and '-'
	const char *whitelist_re = "^snap\\.[a-z][a-z0-9-]*\\.[a-zA-Z0-9-]+$";
	regex_t re;
	if (regcomp(&re, whitelist_re, REG_EXTENDED | REG_NOSUB) != 0)
		die("can not compile regex %s", whitelist_re);

	int status = regexec(&re, appname, 0, NULL, 0);
	regfree(&re);

	return (status == 0);
}

File: snap/naming/snap.go
// -*- Mode: Go; indent-tabs-mode: t -*-

/*
 * Copyright (C) 2026 Canonical Ltd
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package naming

import "strings"

const (
	Core  = InstanceName("core") // or LegacyCore
	Snapd = InstanceName("snapd")
)

// SnapName is the global name of a snap.
type SnapName string

// String returns the snap name as a plain string.
func (n SnapName) String() string {
	return string(n)
}

// InstanceName is the name of a snap decorated with an optional instance key.
type InstanceName string

// NewInstanceName builds an InstanceName from a snap name and an optional
// instance key. When instanceKey is empty the instance name is just the snap
// name.
func NewInstanceName(snapName SnapName, instanceKey string) InstanceName {
	if instanceKey == "" {
		return InstanceName(snapName)
	}
	return InstanceName(string(snapName) + "_" + instanceKey)
}

// SnapName returns the snap name part of the instance name, dropping the
// instance key if any.
func (n InstanceName) SnapName() SnapName {
	snapName, _, _ := strings.Cut(string(n), "_")
	return SnapName(snapName)
}

// InstanceKey returns the instance key part of the instance name, or the empty
// string if there is none.
func (n InstanceName) InstanceKey() string {
	_, instanceKey, _ := strings.Cut(string(n), "_")
	return instanceKey
}

// String returns the instance name as a plain string.
func (n InstanceName) String() string {
	return string(n)
}
